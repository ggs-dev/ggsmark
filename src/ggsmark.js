import { HtmlRenderer, Parser, Node } from 'commonmark'
import DOMPurify from 'dompurify'

export default function (text) {
  let reader = new Parser()
  let writer = new HtmlRenderer()
  let parsed = reader.parse(text)
  var walker = parsed.walker()
  let event, node

  while ((event = walker.next())) {
    node = event.node

    if (
      node.type === 'text' &&
      !!node.parent &&
      node.parent.type === 'paragraph' &&
      node.literal.match(/\:youtube/)
    ) {
      let nestedEvent = walker.next()
      let nestedNode = nestedEvent.node

      // Make sure text nodes are not fragmented
      while (!!nestedNode.prev && nestedNode.type === 'text') {
        nestedNode.prev.literal = nestedNode.prev.literal + nestedNode.literal
        nestedNode.unlink()
        nestedEvent = walker.next()
        nestedNode = nestedEvent.node
      }

      let matchYoutubeExp = /\:(?:youtube)\s(?:https?\:\/\/)?(?:www\.)?(?:youtube\.com\/watch|youtu\.be\/)(?:\?v\=)?([^\s]+)/
      let splitText = node.literal.split(matchYoutubeExp)

      for (let index in splitText) {
        if (index % 2 == 0) {
          let text = new Node('text')
          text.literal = splitText[index]
          node.insertBefore(text)
        } else {
          let div = new Node('custom_block')
          div.onEnter = '<div class="youtube-video">'
          div.onExit = '</div>'
          let iframe = new Node('custom_inline')
          iframe.onEnter = `<iframe src="https://www.youtube.com/embed/${splitText[index]}" type="text/html" frameborder="0">`
          iframe.onExit = '</iframe>'
          div.appendChild(iframe)
          node.insertBefore(div)
        }
      }

      node.unlink()
    }
    else if(
    node.type === 'text' &&
    !!node.parent &&
    node.parent.type === 'paragraph' &&
    node.literal.match(/\:text-center/)
  ) {
    let nestedEvent = walker.next()
    let nestedNode = nestedEvent.node

    // Make sure text nodes are not fragmented
    while (!!nestedNode.prev && nestedNode.type === 'text') {
      nestedNode.prev.literal = nestedNode.prev.literal + nestedNode.literal
      nestedNode.unlink()
      nestedEvent = walker.next()
      nestedNode = nestedEvent.node
    }

    let matchTextCntrExp = (/\:text-center/)
    let splitText = node.literal.split(matchTextCntrExp)

    for (let index in splitText) {
      if (index % 2 == 0) {
        let text = new Node('text')
        text.literal = splitText[index]
        node.insertBefore(text)
      } else {
        let div = new Node('custom_block')
        div.onEnter = `<div style="text-align: center"> ${splitText[index]}`
        div.onExit = '</div>'
        node.insertBefore(div)
      }
    }

    node.unlink()
  }
  else if(
    node.type === 'text' &&
    !!node.parent &&
    node.parent.type === 'paragraph' &&
    node.literal.match(/\:text-right)
  ) {
    let nestedEvent = walker.next()
    let nestedNode = nestedEvent.node

    // Make sure text nodes are not fragmented
    while (!!nestedNode.prev && nestedNode.type === 'text') {
      nestedNode.prev.literal = nestedNode.prev.literal + nestedNode.literal
      nestedNode.unlink()
      nestedEvent = walker.next()
      nestedNode = nestedEvent.node
    }

    let matchTextCntrExp = (/\:text-right/)
    let splitText = node.literal.split(matchTextCntrExp)

    for (let index in splitText) {
      if (index % 2 == 0) {
        let text = new Node('text')
        text.literal = splitText[index]
        node.insertBefore(text)
      } else {
        let div = new Node('custom_block')
        div.onEnter = `<div style="text-align: right"> ${splitText[index]}`
        div.onExit = '</div>'
        node.insertBefore(div)
      }
    }

    node.unlink()
  }
}

  return DOMPurify.sanitize(writer.render(parsed), { ADD_TAGS: ['iframe'] })
}
