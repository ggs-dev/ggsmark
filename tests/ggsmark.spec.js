import ggsmark from '../src/ggsmark'

describe('render centered text', () => {
  test('single line text center', () => {
    let string = ':text-center Test123'
    let result = ggsmark(string)

    expect(result).toMatchSnapshot()
  })

  test('wrap text center', () => {
    let string = `:text-center yeet :text-center`
    let result = ggsmark(string)

    expect(result).toMatchSnapshot()
  })

  test('text center with some text before and after', () => {
    let string = `this should not be in text center :text-center yeet :text-center after text`
    let result = ggsmark(string)

    expect(result).toMatchSnapshot()
  })

  test('text center in text center', () => {
    let string = ':text-center test :text-center wtf :text-center test :text-center'
    let result = ggsmark(string)

    expect(result).toMatchSnapshot()
  })

  test('use bold text', () => {
    let string = '\n\n:text-center\n\n**test** \n:text-center'
    let result = ggsmark(string)

    expect(result).toMatchSnapshot()
  })

  test('use bold text single line', () => {
    let string = ':text-center **bold**'
    let result = ggsmark(string)

    expect(result).toMatchSnapshot()
  })

  test('use heading', () => {
    let string = '\n\n:text-center\n# test\n:text-center'
    let result = ggsmark(string)

    expect(result).toMatchSnapshot()
  })
})

describe('render right allignment text', () => {
  test('text-right', () => {
    let string = ':text-right Test123'
    let result = ggsmark(string)

    expect(result).toMatchSnapshot()
  })
})

describe('render left allignment text', () => {
  test('text-left', () => {
    let string = ':text-left Test123'
    let result = ggsmark(string)

    expect(result).toMatchSnapshot()
  })
})

describe('render soundcloud blocks', () => {
  test('single line', () => {
    // Arrange
    let string =
      ':soundcloud https://soundcloud.com/iamcardib/wap-feat-megan-thee-stallion'

    // Act
    let result = ggsmark(string)

    // Assert
    expect(result).toMatchSnapshot()
  })

  test('content before and after', () => {
    // Arrange
    let string =
      '**bold text** before text :soundcloud https://soundcloud.com/iamcardib/wap-feat-megan-thee-stallion after text **bold**'

    // Act
    let result = ggsmark(string)

    // Assert
    expect(result).toMatchSnapshot()
  })

  test('multiple occurrences', () => {
    // Arrange
    let string =
      ':soundcloud https://soundcloud.com/iamcardib/wap-feat-megan-thee-stallion :soundcloud https://soundcloud.com/iamcardib/wap-feat-megan-thee-stallion :soundcloud https://soundcloud.com/iamcardib/wap-feat-megan-thee-stallion :soundcloud https://soundcloud.com/iamcardib/wap-feat-megan-thee-stallion'

    // Act
    let result = ggsmark(string)

    // Assert
    expect(result).toMatchSnapshot()
  })
})


describe('render youtube blocks', () => {
  test('repeated youtube with text before and after', () => {
    // Arrange
    let string = `
**bold** string before comyoutube :youtube http://www.youtube./watch?v=52c_QSg64fs after youtube :youtube http://www.youtube.com/watch?v=waefawefwaef *italics*
soft new line

new line
    `

    // Act
    let result = ggsmark(string)

    // Assert
    expect(result).toMatchSnapshot()
  })

  test('repeated youtube', () => {
    // Arrange
    let string = `:youtube http://www.youtube.com/watch?v=52c_QSg64fs :youtube http://www.youtube.com/watch?v=52c_QSg64fs :youtube http://www.youshit.com`

    // Act
    let result = ggsmark(string)

    // Assert
    expect(result).toMatchSnapshot()
  })

  test('youtube with timestamp', () => {
    // Arrange
    let string = `:youtube https://youtu.be/IJhgZBn-LHg?t=712`

    // Act
    let result = ggsmark(string)

    // Assert
    expect(result).toMatchSnapshot()
  })
})
