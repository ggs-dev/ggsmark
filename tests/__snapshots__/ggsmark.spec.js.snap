// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render centered text single line text center 1`] = `
"<p>
</p><div style=\\"text-align: center\\">  Test123
</div>
<p></p>
"
`;

exports[`render centered text text center with some text before and after 1`] = `
"<p>this should not be in text center 
</p><div style=\\"text-align: center\\">  yeet 
</div>
<p></p>
"
`;

exports[`render centered text use bold text 1`] = `
"<p>
</p><div style=\\"text-align: center\\"> 
<strong>test</strong>
</div>
<p></p>
"
`;

exports[`render centered text use bold text single line 1`] = `
"<p>
</p><div style=\\"text-align: center\\">  
<strong>bold</strong>
</div>
<p></p>
"
`;

exports[`render centered text use heading 1`] = `
"<p>
</p><div style=\\"text-align: center\\"> 
<h1>test</h1>
</div>
<p></p>
"
`;

exports[`render left allignment text text-left 1`] = `
"<p>
</p><div style=\\"text-align: left\\">  Test123
</div>
<p></p>
"
`;

exports[`render right allignment text text-right 1`] = `
"<p>
</p><div style=\\"text-align: right\\">  Test123
</div>
<p></p>
"
`;

exports[`render soundcloud blocks content before and after 1`] = `
"<p><strong>bold text</strong> before text 
<iframe src=\\"https://w.soundcloud.com/player/?url=https://soundcloud.com/iamcardib/wap-feat-megan-thee-stallion&amp;color=%23ff5500&amp;auto_play=false&amp;hide_related=false&amp;show_comments=true&amp;show_user=true&amp;show_reposts=false&amp;show_teaser=true&amp;visual=true\\" height=\\"300\\" width=\\"100%\\"></iframe>
 after text <strong>bold</strong></p>
"
`;

exports[`render soundcloud blocks multiple occurrences 1`] = `
"<p>
<iframe src=\\"https://w.soundcloud.com/player/?url=https://soundcloud.com/iamcardib/wap-feat-megan-thee-stallion&amp;color=%23ff5500&amp;auto_play=false&amp;hide_related=false&amp;show_comments=true&amp;show_user=true&amp;show_reposts=false&amp;show_teaser=true&amp;visual=true\\" height=\\"300\\" width=\\"100%\\"></iframe>
 
<iframe src=\\"https://w.soundcloud.com/player/?url=https://soundcloud.com/iamcardib/wap-feat-megan-thee-stallion&amp;color=%23ff5500&amp;auto_play=false&amp;hide_related=false&amp;show_comments=true&amp;show_user=true&amp;show_reposts=false&amp;show_teaser=true&amp;visual=true\\" height=\\"300\\" width=\\"100%\\"></iframe>
 
<iframe src=\\"https://w.soundcloud.com/player/?url=https://soundcloud.com/iamcardib/wap-feat-megan-thee-stallion&amp;color=%23ff5500&amp;auto_play=false&amp;hide_related=false&amp;show_comments=true&amp;show_user=true&amp;show_reposts=false&amp;show_teaser=true&amp;visual=true\\" height=\\"300\\" width=\\"100%\\"></iframe>
 
<iframe src=\\"https://w.soundcloud.com/player/?url=https://soundcloud.com/iamcardib/wap-feat-megan-thee-stallion&amp;color=%23ff5500&amp;auto_play=false&amp;hide_related=false&amp;show_comments=true&amp;show_user=true&amp;show_reposts=false&amp;show_teaser=true&amp;visual=true\\" height=\\"300\\" width=\\"100%\\"></iframe>
</p>
"
`;

exports[`render soundcloud blocks single line 1`] = `
"<p>
<iframe src=\\"https://w.soundcloud.com/player/?url=https://soundcloud.com/iamcardib/wap-feat-megan-thee-stallion&amp;color=%23ff5500&amp;auto_play=false&amp;hide_related=false&amp;show_comments=true&amp;show_user=true&amp;show_reposts=false&amp;show_teaser=true&amp;visual=true\\" height=\\"300\\" width=\\"100%\\"></iframe>
</p>
"
`;

exports[`render youtube blocks repeated youtube 1`] = `
"<p>
</p><div class=\\"ggsmark-youtube\\">
<iframe type=\\"text/html\\" src=\\"https://www.youtube.com/embed/\${splitText[index]}\\"></iframe>
</div>
 
<div class=\\"ggsmark-youtube\\">
<iframe type=\\"text/html\\" src=\\"https://www.youtube.com/embed/\${splitText[index]}\\"></iframe>
</div>
 :youtube http://www.youshit.com<p></p>
"
`;

exports[`render youtube blocks repeated youtube with text before and after 1`] = `
"<p><strong>bold</strong> string before comyoutube :youtube http://www.youtube./watch?v=52c_QSg64fs after youtube 
</p><div class=\\"ggsmark-youtube\\">
<iframe type=\\"text/html\\" src=\\"https://www.youtube.com/embed/\${splitText[index]}\\"></iframe>
</div>
 <em>italics</em>
soft new line<p></p>
<p>new line</p>
"
`;

exports[`render youtube blocks youtube with timestamp 1`] = `
"<p>
</p><div class=\\"ggsmark-youtube\\">
<iframe type=\\"text/html\\" src=\\"https://www.youtube.com/embed/\${splitText[index]}\\"></iframe>
</div>
<p></p>
"
`;
